src/Concordium/Client/GRPC2.hs

```diff
@@ -0,0 +1,2766 @@
+    fromProto = return . IP_ID . deMkWord32
+
+instance FromProto Proto.CredentialPublicKeys where
+    type Output' Proto.CredentialPublicKeys = CredentialPublicKeys
+    fromProto cpk = do
+        credKeys <-
+            fmap Map.fromAscList
+                . mapM convert
+                . Map.toAscList
+                =<< cpk ^? ProtoFields.keys
+        credThreshold <-
+            fmap SignatureThreshold
+                . deMkWord8
+                =<< cpk ^? ProtoFields.threshold
+        -- This can not be 0.
+        if credThreshold == 0
+            then Nothing
+            else return CredentialPublicKeys{..}
+      where
+        convert (ki, pKey) = do
+            key <- pKey ^? ProtoFields.ed25519Key
+            k <- case S.decode key of
+                Left _ -> Nothing
+                Right k -> return k
+            -- Ensure that the index fits in a Word8.
+            if fromIntegral ki > (maxBound :: Word8)
+                then Nothing
+                else return (fromIntegral ki, VerifyKeyEd25519 k)
+
+instance FromProto Proto.AccountInfo where
+    type Output' Proto.AccountInfo = AccountInfo
+    fromProto ai = do
+        aiAccountNonce <- fromProto =<< ai ^? ProtoFields.sequenceNumber
+        aiAccountAmount <- fromProto =<< ai ^? ProtoFields.amount
+        aiAccountReleaseSchedule <- fromProto =<< ai ^? ProtoFields.schedule
+        aiAccountCredentials <-
+            do
+                fmap Map.fromAscList
+                . mapM convert
+                . Map.toAscList
+                =<< ai ^? ProtoFields.creds
+        aiAccountThreshold <- fromProto =<< ai ^? ProtoFields.threshold
+        aiAccountEncryptedAmount <- fromProto =<< ai ^? ProtoFields.encryptedBalance
+        aiAccountEncryptionKey <- fromProto =<< ai ^? ProtoFields.encryptionKey
+        aiAccountIndex <- fromProto =<< ai ^? ProtoFields.index
+        aiStakingInfo <-
+            case ai ^. ProtoFields.maybe'stake of
+                Nothing -> return AccountStakingNone
+                Just asi -> fromProto asi
+        aiAccountAddress <- fromProto =<< ai ^? ProtoFields.address
+        return AccountInfo{..}
+      where
+        versionTag = 0
+        convert (key, val) = do
+            v <- fromProto val
+            -- Ensure that the index fits in a Word8.
+            if fromIntegral key > (maxBound :: Word8)
+                then Nothing
+                else return (CredentialIndex $ fromIntegral key, Versioned versionTag v)
+
+instance ToProto BlockHashInput where
+    type Output BlockHashInput = Proto.BlockHashInput
+    toProto = \case
+        Best -> Proto.make $ ProtoFields.best .= defMessage
+        LastFinal -> Proto.make $ ProtoFields.lastFinal .= defMessage
+        Given bh -> Proto.make $ ProtoFields.given .= toProto bh
+
+instance FromProto Proto.BlockHash where
+    type Output' Proto.BlockHash = BlockHash
+    fromProto bh = BlockHash <$> deMkSerialize bh
```

Comment:
Yes, as mentioned `S.decode` fails. I tested it with `getBlocks`.

Reply:
What is the raw value you are converting?

Reply:
Here are some examples of the received blockhash bytestrings dumped as hex/escaped octal(?):

```
Hex: \"3199f6f27393783c7fb970054b39a0d9b3177c3e605a66931b8fbbee9141db5f\"
Raw: 1\153\246\242s\147x<\DEL\185p\ENQK9\160\217\179\ETB|>`Zf\147\ESC\143\187\238\145A\219_

Hex: \"f95cb13af02e9a1275fa6a0a58833a8eb304d0784097e7e17eab9d10cd1901ea\"
Raw: \249\\\177:\240.\154\DC2u\250j\nX\131:\142\179\EOT\208x@\151\231\225~\171\157\DLE\205\EM\SOH\234

Hex: \"64b09e4d666600907ab37756428e58664122e5bddd0d738f12acf1ee20a024bf\"
Raw: d\176\158Mff\NUL\144z\179wVB\142XfA\"\229\189\221\rs\143\DC2\172\241\238 \160$\191

Hex: \"1cd7b1debe02715a09a629b338643b1bdd63c649570771e8289ba7391388ecd8\"
Raw: \FS\215\177\222\190\STXqZ\t\166)\179\&8d;\ESC\221c\198IW\aq\232(\155\167\&9\DC3\136\236\216

Hex: \"813b728c41921b207b79cfc5bdea2db9a933d846228ec0671f4ce447906b71a4\"
Raw: \129;r\140A\146\ESC {y\207\197\189\234-\185\169\&3\216F\"\142\192g\USL\228G\144kq\164
```

Reply:
Are you sure that it is this conversion that does not work?

Because
```
$ Data.Serialize.decode (fromString "d\176\158Mff\NUL\144z\179wVB\142XfA\"\229\189\221\rs\143\DC2\172\241\238 \160$\191") :: Either String Concordium.Crypto.SHA256.Hash
Right 64b09e4d666600907ab37756428e58664122e5bddd0d738f12acf1ee20a024bf
```

Reply:
Of course this is due to a bug I have in`deMkSerialize`, it does not wrap the return value in the second case:

```
deMkSerialize val = case S.decode $ val ^. ProtoFields.value of
    Left _ -> Nothing
    Right v -> v
```

Sorry about the the confusion.

Reply:
Another instance of "code that sufficently generic in types is very similar to untyped code" :)